### ISSUE TITLE
Can't copy a file from remote to host

### ISSUE THREAD

**Issue by rblaack** [FreeRDP/FreeRDP#7904]
**Found a bug? - We would like to help you and smash the bug away.**
1. __Please don't "report" questions as bugs. For these (questions/build instructions/...) please use one of the following means of contact:__
   * We are reachable via:
    * Matrix room : #FreeRDP:matrix.org (main)
    * XMPP channel: #FreeRDP#matrix.org@matrix.org (bridged)
    * IRC channel : #freerdp @ irc.oftc.net (bridged)
   * We are reachable via mailing list <freerdp-devel@lists.sourceforge.net>
   * Try our mailing list for discussions/questions
1. Before reporting a bug have a look into our issue tracker to see if the bug was already reported and you can add some additional information.
1. If it's a __new__ bug - create a new issue.
1. For more details see https://github.com/FreeRDP/FreeRDP/wiki/BugReporting

**Describe the bug**
Copy a file from remote to desktop is not working, however sending from host to remote is working. I checked previous issue and this one should be fixed since this feature has been implemented to master. #6699 

**To Reproduce**
1. Go to remote server and copy a file.
2. Past to user directory on host.
3. Error during copy : [10:44:28:466] [18905:18930] [WARN][com.freerdp.client.x11] - failed to get clipboard data in format text/uri-list [source format FileGroupDescriptorW]

**Expected behavior**
The file should be able to be copied from remote and be past in host.

**Screenshots**
![image](https://user-images.githubusercontent.com/85481937/168554553-3749fa57-00b4-4a53-bbc9-ab9c4f8216f7.png)

**Application details**
* FreeRDP version : FreeRDP version 2.7.0 (2.7.0)
* Command line used : WLOG_LEVEL=DEBUG xfreerdp +clipboard /dynamic-resolution /v:10.10.52.126 /u:auser
* Output of `xfreerdp /buildconfig` : `Build configuration: BUILD_TESTING=OFF BUILTIN_CHANNELS=ON HAVE_AIO_H=1 HAVE_EXECINFO_H=1 HAVE_FCNTL_H=1 HAVE_GETLOGIN_R=1 HAVE_GETPWUID_R=1 HAVE_INTTYPES_H=1 HAVE_JOURNALD_H=TRUE HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK=ON HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIBS= HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL=1 HAVE_SYSLOG_H=1 HAVE_SYS_EVENTFD_H=1 HAVE_SYS_FILIO_H= HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H= HAVE_SYS_STRTIO_H= HAVE_SYS_TIMERFD_H=1 HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 HAVE_XI_TOUCH_CLASS=1 WITH_ALSA=ON WITH_CAIRO=OFF WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CLANG_FORMAT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUPS=ON WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_RDPGFX=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF_AVAILABLE=0 WITH_DEBUG_URBDRC=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_DSP_FFMPEG=ON WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FFMPEG=TRUE WITH_FFMPEG=TRUE WITH_GFX_H264=ON WITH_GPROF=OFF WITH_GSM=OFF WITH_GSSAPI=OFF WITH_ICU=ON WITH_IPP=OFF WITH_JPEG=ON WITH_LAME=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=ON WITH_MACAUDIO=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_OPENCL=OFF WITH_OPENH264=OFF WITH_OPENSLES=OFF WITH_OPENSSL=ON WITH_OSS=ON WITH_PAM=ON WITH_PCSC=ON WITH_PROFILER=OFF WITH_PROXY=ON WITH_PROXY_MODULES=OFF WITH_PULSE=ON WITH_SAMPLE=OFF WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SERVER=ON WITH_SERVER_CHANNELS=ON WITH_SERVER_INTERFACE=ON WITH_SHADOW=ON WITH_SMARTCARD_INSPECT=OFF WITH_SOXR=OFF WITH_SSE2=ON WITH_SWSCALE=ON WITH_THIRD_PARTY=OFF WITH_VAAPI=OFF WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_VERBOSE_WINPR_ASSERT=ON WITH_WAYLAND=ON WITH_WINPR_TOOLS=ON WITH_X11=ON WITH_XCURSOR=ON WITH_XDAMAGE=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XKBFILE=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XSHM=ON WITH_XTEST=ON WITH_XV=ON WITH_ZLIB=ON
Build type:          None
CFLAGS:              -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -flto=auto -fPIC -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -g -fno-omit-frame-pointer -DWINPR_DLL
Compiler:            GNU, 11.2.0
Target architecture: x64
`
* OS version connecting to (server side) : Windows 11 / Windows Server 2019

**Environment (please complete the following information):**
 - OS: Linux
 - Version/Distribution: ArchLinux ( Kernel 5.17.5-arch1-2)
 - Architecture: [amd64, arm]: amd64
 - Graphic Platform: KDE (5.24.5) X11

**Comment by pnowack** [comment]
I don't use KDE, so not sure why this is failing here. It works here on GNOME (which uses the same original function, but with different arguments). There are a lot of possibilities here, where things can fail.

However, on a brief look, something looks off to me:
```c
/* Prepend header of kde dolphin format to file list*/
static void* convert_filedescriptors_to_uri_list(wClipboard* clipboard, UINT32 formatId,
                                                 const void* data, UINT32* pSize)
{
	return convert_filedescriptors_to_file_list(clipboard, formatId, data, pSize, "", "file:", "\n",
	                                            FALSE);
}
```
This above, is the current code here for `text/uri-list`, but it should be:
```c
/* Prepend header of kde dolphin format to file list*/
static void* convert_filedescriptors_to_uri_list(wClipboard* clipboard, UINT32 formatId,
                                                 const void* data, UINT32* pSize)
{
	return convert_filedescriptors_to_file_list(clipboard, formatId, data, pSize, "", "file://", "\r\n",
	                                            FALSE);
}
```

I don't think it would solve this issue, as apparently it is failing on your side already. However, these lines need to be changed too here, I think.

**Comment by pnowack** [comment]
> i wonder if it working before on KDE ?

Don't know. But I think the `convert_filedescriptors_to_file_list ()` call is failing here. It would need testing, at which condition `NULL` is returned.  
The issue might be related to the file names here. If there is a character, that needs escaping, then this is an indicator here.
So, for testing, would be good if you can try copy pasting files with file names, containing only the usual english alphabet.
If this works, then an example file list needs to be known as test case, to find the characters, that cause the issue here (likely need some escaping here, if that is the issue here).

**Comment by akallabeth** [comment]
actually it is quite simple, the support for file copy with fuse is currently only available on master.
stable-2.0 does not support that yet.

**Comment by akallabeth** [comment]
@pnowack from a fast look the `//` is part of what is expected from remote, but don´t have time for a thorough check.

**Comment by KangLin** [comment]
> I don't use KDE, so not sure why this is failing here. It works here on GNOME (which uses the same original function, but with different arguments). There are a lot of possibilities here, where things can fail.
> 
> However, on a brief look, something looks off to me:
> 
> ```c
> /* Prepend header of kde dolphin format to file list*/
> static void* convert_filedescriptors_to_uri_list(wClipboard* clipboard, UINT32 formatId,
>                                                  const void* data, UINT32* pSize)
> {
> 	return convert_filedescriptors_to_file_list(clipboard, formatId, data, pSize, "", "file:", "\n",
> 	                                            FALSE);
> }
> ```
> 
> This above, is the current code here for `text/uri-list`, but it should be:
> 
> ```c
> /* Prepend header of kde dolphin format to file list*/
> static void* convert_filedescriptors_to_uri_list(wClipboard* clipboard, UINT32 formatId,
>                                                  const void* data, UINT32* pSize)
> {
> 	return convert_filedescriptors_to_file_list(clipboard, formatId, data, pSize, "", "file://", "\r\n",
> 	                                            FALSE);
> }
> ```
> 
> I don't think it would solve this issue, as apparently it is failing on your side already. However, these lines need to be changed too here, I think.

See:
- GTK: https://docs.gtk.org/glib/struct.Uri.html
- uri syntax: https://www.rfc-editor.org/rfc/rfc3986#section-3
- uri-lists format: https://www.rfc-editor.org/rfc/rfc2483#section-5

Obviously the [current code](https://github.com/FreeRDP/FreeRDP/blob/fd96ed0acd3af556340943d569e4738a87e63dd2/winpr/libwinpr/clipboard/posix.c#L779) it is wrong.
[Current comment](https://github.com/FreeRDP/FreeRDP/blob/fd96ed0acd3af556340943d569e4738a87e63dd2/winpr/libwinpr/clipboard/posix.c#L853) is also wrong.

[My program](https://github.com/KangLin/RabbitRemoteControl/tree/develop) is write qt. I test current code is wrong. use follow code is successed:

 /* Prepend header of kde dolphin format to file list*/
static void* convert_filedescriptors_to_uri_list(wClipboard* clipboard, UINT32 formatId,
                                                  const void* data, UINT32* pSize)
 {
 	return convert_filedescriptors_to_file_list(clipboard, formatId, data, pSize, "", "file://", "\r\n",
 	                                            FALSE);
 }

I modified in #8187.

@jackyzy823 You are committed in 2e0b5e07. What is your basis for modification?

**Comment by KangLin** [comment]
uri-lists format: https://www.rfc-editor.org/rfc/rfc2483#section-5

The format of text/uri-list resources is:

    1) Any lines beginning with the '#' character are comment lines
        and are ignored during processing. (Note that URIs may contain
        the '#' character, so it is only a comment character when it is
        the first character on a line.)

    2) The remaining non-comment lines shall be URIs (URNs or URLs),
        encoded according to the URL or URN specifications ([RFC2141](https://www.rfc-editor.org/rfc/rfc2141),
        [RFC1738](https://www.rfc-editor.org/rfc/rfc1738) and [RFC2396](https://www.rfc-editor.org/rfc/rfc2396)). Each URI shall appear on one and only one
        line. Very long URIs are not broken in the text/uri-list format.
        Content-transfer-encodings may be used to enforce line length
        limitations.

    3) As for all text/* formats, lines are terminated with a CRLF pair.

**Comment by jackyzy823** [comment]
Well .... maybe i wrote wrong comment? but maybe i wrote proper code
https://github.com/FreeRDP/FreeRDP/pull/6699/files#diff-59cb029fe0e9ac166994e1bf59e3414c58d76c076553a6484d95b72b474e81afR755

And Qt may ignore '\r' when parsing. see https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qmimedata.cpp?h=dev#n156
so maybe it's safe to just use '\n'

**Comment by KangLin** [comment]
@jackyzy823 Yes, your code is correct. but @akallabeth has been modified it in 453d3b4c8b5b1d0cf4d36d0ea1f5008cf7cdee7e . I see #7754 . should be modified `x-special/gnome-copied-files`, not text/uri-list .

@pnowack has been modified in two places:
file:// and \r\n

I have tested on Qt that file:// is wrong too.

Current issuses is in KDE. please @rblaack test https://github.com/FreeRDP/FreeRDP/pull/8187 or https://github.com/KangLin/FreeRDP/tree/uri-list

**Comment by pnowack** [comment]
> And Qt may ignore '\r' when parsing. see https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qmimedata.cpp?h=dev#n156
so maybe it's safe to just use '\n'

I don't see much of a reason to violate the spec here. Every uri MUST be ended with CRLF (`\r\n`): https://www.rfc-editor.org/rfc/rfc2483#section-5 (`As for all text/* formats, lines are terminated with a CRLF pair.`).

> I have tested on Qt that file:// is wrong too.

No, `file://` needs to prepend our uris in the lists. See also https://datatracker.ietf.org/doc/html/rfc8089#appendix-B

Example:

Assuming the fuse path is `/path/to/fuseDir`, and the on the other peer the files and folders `a`, `a\\b.txt`, `c.bmp`, then our `text/uri-list` is:

```
file:///path/to/fuseDir/a\r\n
file:///path/to/fuseDir/a/b.txt\r\n
file:///path/to/fuseDir/c.bmp\r\n
```

**Comment by jackyzy823** [comment]
> No, `file://` needs to prepend our uris in the lists. See also https://datatracker.ietf.org/doc/html/rfc8089#appendix-B

Sometimes (historically) KDE do not follow RFCs

See complains from https://phpspot.net/php/man/gtk/tutorials.filednd.urilist.html
> KDE programs like Konqueror, Kate and so pass the following: file:/path/to/file. Only one "/", totally incorrect. 


Anyway. I found that KDE have ability to automatically convert malformed uri-list (like no / or \n) to correct one.

**Comment by KangLin** [comment]
@jackyzy823 @pnowack
Whether or not other software converts \n to \r\n correctly. We MUST follow the RFC.


About URI:
Please see https://datatracker.ietf.org/doc/html/rfc8089#appendix-E.2 and https://datatracker.ietf.org/doc/html/rfc8089#appendix-F



   o  A traditional file URI for a local file with an empty authority.
      For example:

      *  "file:///path/to/file"

   o  The minimal representation of a local file with no authority field
      and an absolute path that begins with a slash "/".  For example:

      *  "file:/path/to/file"

   o  The minimal representation of a local file in a DOS- or Windows-
      based environment with no authority field and an absolute path
      that begins with a drive letter.  For example:

      *  "file:c:/path/to/file"

   o  Regular DOS or Windows file URIs with vertical line characters in
      the drive letter construct.  For example:

      *  "file:///c|/path/to/file"

      *  "file:/c|/path/to/file"

      *  "file:c|/path/to/file"


If only the prefix file:// is added, the following error format will be generated under Windows:

    file://c:path/to/file


To sum up, adding the prefix "file:" is correct.
If you want to prefix "file://" then also check if the first character of Path is "/" .

**Comment by jackyzy823** [comment]
First of all. RFCs are not Standards. 

For `file:`
In this situation (implmenting copy file from remote to local)  `uri-list` is only  used under POSIX system to provide PATH like `/tmp/.xfreerdp.xxx/`  , not related with Windows.

So i think both are acceptable , one means `empty authority` . the other means `no authority field`

For CRLF
the rfc says
 > As for all text/* formats, lines are terminated with a CRLF pair

It may not mean **each** line . It could mean like `a\nb\nc\r\n`

**Comment by KangLin** [comment]
> First of all. RFCs are not Standards.

What is standards?

> For `file:` In this situation (implmenting copy file from remote to local) `uri-list` is only used under POSIX system to provide PATH like `/tmp/.xfreerdp.xxx/` , not related with Windows.
> 
> So i think both are acceptable , one means `empty authority` . the other means `no authority field`

 [My program](https://github.com/KangLin/RabbitRemoteControl/tree/develop) is used the situation. It is run in multi-os(include Windows).


> For CRLF the rfc says
> 
> > As for all text/* formats, lines are terminated with a CRLF pair
> 
> It may not mean **each** line . It could mean like `a\nb\nc\r\n`

Please understand in context.

**Comment by jackyzy823** [comment]
> [My program](https://github.com/KangLin/RabbitRemoteControl/tree/develop) is used the situation. It is run in multi-os(include Windows).

This code only used on xfreerdp client  , so do xfreerdp client run on Windows ?

The wfreerdp (windows version freerdp client) need not convert to uri-list. They just use file descriptor.


> Please understand in context.

This requirement is applied to  **ALL** text/* format. so do every format listed in [HERE](https://www.iana.org/assignments/media-types/media-types.xhtml#text) (like text/html  etc ) needs CRLF each line?

**Comment by KangLin** [comment]
> > [My program](https://github.com/KangLin/RabbitRemoteControl/tree/develop) is used the situation. It is run in multi-os(include Windows).
> 
> This code only used on xfreerdp client , so do xfreerdp client run on Windows ?
> 
> The wfreerdp (windows version freerdp client) need not convert to uri-list. They just use file descriptor.
> 

I'm modified it. Please see: https://github.com/KangLin/FreeRDP/tree/stable-2.0
or bd275b2fa1b411e854947f64143a934d456df32c

**Comment by pnowack** [comment]
> I'm modified it. Please see: https://github.com/KangLin/FreeRDP/tree/stable-2.0
or https://github.com/FreeRDP/FreeRDP/commit/bd275b2fa1b411e854947f64143a934d456df32c

??? As written above in https://github.com/FreeRDP/FreeRDP/issues/7904#issuecomment-1242882834, MS Windows uses `FileGroupDescriptorW` on its site for copy pasting files and folders via the clipboard.
On the linux side we cannot do this directly. So we use FUSE for the filesystem, create uris that point to our FUSE filesystem and when a file is read from our FUSE filesystem, the request is forwarded to the other peer.
Since MS Windows works here completely different (there is no FUSE on Windows and it is not even needed there), I fail to see the rationale here.

**Comment by KangLin** [comment]
Because I'm using qt, I need this part to work under Windows as well. 
The current practice is to first copy the peer file to the local, and then return the local file name.
In the future I will wrap a platform independent stream object on top of fuse and FileGroupDescriptorW and FileContent to handle.

**Comment by KangLin** [comment]
Please @pnowack  @jackyzy823 review: https://github.com/FreeRDP/FreeRDP/pull/8212
